landlab/graph/framed_voronoi/framed_voronoi.py

```diff
@@ -0,0 +1,474 @@
+            When set, controls a pseudo-randomness of moves to ensure
+            reproducibility.
+            When None, seed is random and the moves of coordinates are
+            completely random.
+
+        Returns
+        -------
+        x_of_node, y_of_node : ndarray of float
+            The arrays of x and y coordinates.
+
+        Examples
+        --------
+        >>> from landlab.graph.framed_voronoi.framed_voronoi import HorizontalRectVoronoiGraph
+        >>> HorizontalRectVoronoiGraph.xy_of_node((3, 3), seed=(200, 500))[0][3]       # doctest: +NORMALIZE_WHITESPACE
+        0.0
+        >>> HorizontalRectVoronoiGraph.xy_of_node((3, 3), seed=(200, 500))[0][5]       # doctest: +NORMALIZE_WHITESPACE
+        2.0
+        >>> HorizontalRectVoronoiGraph.xy_of_node((3, 3), seed=(200, 500))[1][3]       # doctest: +NORMALIZE_WHITESPACE
+        0.749
+        """
+
+        n_rows, n_cols = shape
+        max_move = [
+            (xy_spacing[0] - xy_min_spacing[0]) / 2,
+            (xy_spacing[1] - xy_min_spacing[1]) / 2,
+        ]
+        n_nodes = HorizontalRectVoronoiGraph.number_of_nodes(shape)
+        n_core_nodes = n_nodes - HorizontalRectVoronoiGraph.number_of_perimeter_nodes(
+            shape
+        )
+
+        # Generation of a rectangular grid, coordinates must be float
+        x_of_node, y_of_node = np.meshgrid(
+            np.arange(n_cols, dtype=float) * xy_spacing[0] + xy_of_lower_left[0],
+            np.arange(n_rows, dtype=float) * xy_spacing[1] + xy_of_lower_left[1],
+        )
+        # Randomly move the coordinates of the core nodes of the grid. Move below +/- (spacing - min_spacing)/2
+        if seed is None:
+            xy_random_generator = (np.random.default_rng(), np.random.default_rng())
+        else:
+            xy_random_generator = (
+                np.random.Generator(np.random.PCG64(seed[0])),
+                np.random.Generator(np.random.PCG64(seed[1])),
+            )
+
+        x_y_random_move = (
+            xy_random_generator[0].random(n_core_nodes) * max_move[0] * 2 - max_move[0],
+            xy_random_generator[1].random(n_core_nodes) * max_move[1] * 2 - max_move[1],
+        )
+        if n_rows > 2 and n_cols > 2:
+            x_of_node[1:-1, 1:-1] += np.reshape(
+                x_y_random_move[0], (n_rows - 2, n_cols - 2)
+            )
+            y_of_node[1:-1, 1:-1] += np.reshape(
+                x_y_random_move[1], (n_rows - 2, n_cols - 2)
+            )
+        # Control the node id attribution for left and right edge. For instance, for a 3x3 grid,
+        # make sure that node 3 is at the left of the 2nd row and node 5 at the right.
+        # For this, for each core row, set y of the leftmost node as the minimal y of the row
+        # and set y of the rightmost node as the maximal y of the row
+        for i in range(1, n_rows - 1):
+            y_of_node[i, 0] -= max_move[1] + 1.0e-3
+            y_of_node[i, n_cols - 1] += max_move[1] + 1.0e-3
```

Comment:
@sebastien-lenard I believe you do this only so you can easily get the perimeter nodes, is that correct? It may not matter for what you are using this for but it seems like it may introduce some bias to the node layout that may matter in some situations.

I've "fixed" this (i.e. all edge nodes are not moved from their initial position) on my branch and was thinking I would create a pull request with those changes after this pull request is merged. That way you can review what I've done before incorporating it.

Reply:
Not only I want the list of perimeter nodes, but I also want to know what is the node id at the right of the grid and what is the node id at the left, at a certain "row").